ranger-plugin/service/src/main/java/org/apache/hawq/ranger/authorization/RangerHawqAuthorizer.java

```diff
@@ -0,0 +1,263 @@
+        }
+
+        return response;
+    }
+
+    /**
+     * Authorizes access to a single resource for a given user.
+     *
+     * @param resourceAccess resource to authorize access to
+     * @param user user requesting authorization
+     * @return true if access is authorized, false otherwise
+     */
+    private boolean authorizeResource(ResourceAccess resourceAccess, String user) {
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format("Request: access for user=%s to resource=%s with privileges=%s",
+                user, resourceAccess.getResource(), resourceAccess.getPrivileges()));
+        }
+
+        RangerAccessResourceImpl rangerResource = new RangerAccessResourceImpl();
+        //resource.setOwnerUser();
+        for (Map.Entry<HawqResource, String> resourceEntry : resourceAccess.getResource().entrySet()) {
+            rangerResource.setValue(resourceEntry.getKey().name(), resourceEntry.getValue());
+        }
+
+        boolean accessAllowed = true;
+        // iterate over all privileges requested
+        for (HawqPrivilege privilege : resourceAccess.getPrivileges()) {
+            // TODO not clear how we will get user groups -- Kerberos case ?
+            Set<String> userGroups = Collections.emptySet();
+            boolean privilegeAuthorized = authorizeResourcePrivilege(rangerResource, privilege.name(), user, userGroups);
+            // ALL model of evaluation -- all privileges must be authorized for access to be allowed
+            if (!privilegeAuthorized) {
+                accessAllowed = false;
+                break; // terminate early if even a single privilege is not authorized
+            }
+        }
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug(String.format("Decision: accessAllowed=%s for user=%s to resource=%s with privileges=%s",
+                    accessAllowed, user, resourceAccess.getResource(), resourceAccess.getPrivileges()));
+        }
+
+        return accessAllowed;
+    }
+
+    /**
+     * Authorizes access of a given type (privilege) to a single resource for a given user.
+     *
+     * @param rangerResource resource to authorize access to
+     * @param accessType privilege requested for a given resource
+     * @param user user requesting authorization
+     * @param userGroups groups a user belongs to
+     * @return true if access is authorized, false otherwise
+     */
+    private boolean authorizeResourcePrivilege(RangerAccessResource rangerResource, String accessType, String user, Set<String> userGroups) {
+
+        Map<String, String> resourceMap = rangerResource.getAsMap();
+        String database = resourceMap.get(HawqResource.database.name());
+        String schema = resourceMap.get(HawqResource.schema.name());
+        int resourceSize = resourceMap.size();
+
+        // special handling for non-leaf policies
+        if (accessType.equals(HawqPrivilege.create.name()) && database != null && schema == null && resourceSize == 1) {
```

Comment:
What's the meaning of resourceSize valued as 1 and 2? Could you add some comments in the code here? Thanks